server/app/auth/oidc/OidcProvider.java

```diff
@@ -0,0 +1,185 @@
+
+  private String providerNameConfigName = "provider_name";
+  private String clientIDConfigName = "client_id";
+  private String clientSecretConfigName = "client_secret";
+  private String discoveryURIConfigName = "discovery_uri";
+  private String responseModeConfigName = "response_mode";
+  private String extraScopesConfigName = "additional_scopes";
+
+  public OidcProvider(
+      Config configuration,
+      ProfileFactory profileFactory,
+      Provider<UserRepository> applicantRepositoryProvider) {
+    this.configuration = checkNotNull(configuration);
+    this.profileFactory = checkNotNull(profileFactory);
+    this.applicantRepositoryProvider = applicantRepositoryProvider;
+
+    baseUrl = getConfigurationValue("base_url");
+  }
+
+  /*
+   * Provide the prefix used in the application.conf
+   */
+  protected abstract String attributePrefix();
+
+  /*
+   * Provide the profile adaptor that should be used.
+   */
+  public abstract ProfileCreator getProfileAdapter(OidcConfiguration config, OidcClient client);
+
+  protected String getConfigurationValue(String attr, String defaultValue) {
+    if (configuration.hasPath(attr)) {
+      return configuration.getString(attr);
+    }
+    return defaultValue;
+  }
+
+  protected String getConfigurationValue(String attr) {
+    return getConfigurationValue(attr, "");
+  }
+
+  protected String getProviderName() {
+    return getConfigurationValue(attributePrefix() + "." + providerNameConfigName);
+  }
+
+  protected String getClientID() {
+    return getConfigurationValue(attributePrefix() + "." + clientIDConfigName);
+  }
+
+  protected String getClientSecret() {
+    return getConfigurationValue(attributePrefix() + "." + clientSecretConfigName);
+  }
+
+  protected String getDiscoveryURI() {
+    return getConfigurationValue(attributePrefix() + "." + discoveryURIConfigName);
+  }
+
+  protected String getResponseMode() {
+    return getConfigurationValue(attributePrefix() + "." + responseModeConfigName);
+  }
+
+  protected String getResponseType() {
+    // Our local fake IDCS doesn't support 'token' auth.
+    if (baseUrl.contains("localhost:")) {
+      return "id_token";
+    }
+    return "id_token token";
+  }
+
+  protected String[] getExtraScopes() {
+    return getConfigurationValue(attributePrefix() + "." + extraScopesConfigName).split(" ");
+  }
+
+  protected String getCallbackURL() {
+    return baseUrl + "/callback";
+  }
+
+  protected String getScopes() {
+    // Scopes are the other things that we want from the OIDC endpoint
+    // (needs to also be configured on provider side).
+    String[] extraScopes = getExtraScopes();
+
+    ArrayList<String> allClaims = new ArrayList<>();
+    Collections.addAll(allClaims, defaultScopes);
+    Collections.addAll(allClaims, extraScopes);
+    return String.join(" ", allClaims);
+  }
+
+  @Override
+  public OidcClient get() {
+    String clientID = Objects.toString(getClientID(), "");
```

Comment:
why do you need Objects.toString here? your methods already return Strings.

Reply:
It provides the null default.  A class that inherits from this can overrite these, and return a null.